README.md
# animated_grid

Animated grid view.
Currently, it is only available in page mode.
Scrolling to switch pages is not supported.

![OverView](https://github.com/suzukihund/animated_grid/blob/master/sample_gif/overview.gif?raw=true)

## Getting Started

In the `pubspec.yaml` of your flutter project, add the following dependency:
```yaml
animated_grid:
```
In your library add the following import:

```dart
import 'package:animated_grid/animated_grid.dart';
```

Pass an array of key values that identify the elements of the grid. 
It animates the elements in this array as they change.
```dart
var _keys = [1, 2, 3];
AnimatedGrid(
    keys: _keys,
    width: MediaQuery.of(context).size.width,
    height: MediaQuery.of(context).size.height - 44,
    cellRowNum: 2,
    cellColNum: 4,
    sortOrder: SortOrder.lightToLeft,
    scrollDirection: Axis.horizontal,
    onPageChanged: (page) {
      print(page);
    },
    builder: (ctx, index, _) {
      return ExampleContent(
        caption: '${_keys[index]}',
        keyVal: _keys[index],
        onDeleteAt: (key) {
          setState(() {
            _keys.remove(key);
          });
        },
      );
    },
)
```


CHANGELOG.md
## [0.3.0-nullsafety.0]

* support null safety

## [0.2.4]

* modify slide animations

## [0.2.3]

* fix bugs

## [0.2.2]

* update builder callback

## [0.2.1]

* fix typo

## [0.2.0]

* Slide animations are no longer run when transitioning from one page to the next.
* Add PageController parameter.

## [0.1.0] 

* Initial Open Source release.



test/animated_grid_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter/material.dart';
import 'package:animated_grid/animated_grid.dart';

class ExampleApp extends StatefulWidget {
  @override
  _ExampleAppState createState() => _ExampleAppState();
}

class _ExampleAppState extends State<ExampleApp> {
  final _keys = <int>[];
  var _count = 0;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      color: Colors.black,
      home: Builder(
        builder: (context) {
          return Scaffold(
            appBar: AppBar(
                title: Text('Example'),
                actions: [
                  IconButton(
                    key: Key('plus'),
                    icon: Icon(Icons.add),
                    onPressed: () {
                      setState(() {
                        _keys.add(_count);
                        _count++;
                      });
                    },
                  ),
                ]
            ),
            backgroundColor: Colors.black,
            body: AnimatedGrid(
              keys: _keys,
              width: MediaQuery.of(context).size.width,
              height: MediaQuery.of(context).size.height - 44,
              cellRowNum: 2,
              cellColNum: 2,
              sortOrder: SortOrder.topToBottom,
              scrollDirection: Axis.vertical,
              onPageChanged: (page) {
                print(page);
              },
              builder: (ctx, index, _) {
                return ExampleContent(
                  caption: '${_keys[index]}',
                  keyVal: _keys[index],
                  onDeleteAt: (key) {
                    setState(() {
                      _keys.remove(key);
                    });
                  },
                );
              },
            ),
          );
        },
      ),
    );
  }
}

class ExampleContent extends StatelessWidget {
  const ExampleContent({Key? key, this.caption, this.keyVal, this.onDeleteAt}) : super(key: key);

  final String? caption;
  final int? keyVal;
  final Function(int?)? onDeleteAt;

  @override
  Widget build(BuildContext context) {
    return Container(
      child: Column(
        children: [
          Text(caption!),
          FlatButton(
            key: Key(caption!),
            onPressed: () {
              print('deleted $caption');
              onDeleteAt!(keyVal);
            },
            child: Text('Delete $caption'),
          ),
        ],
      ),
    );
  }
}

void main() {
  testWidgets('add cell and delete cell', (tester) async {
    await tester.pumpWidget(ExampleApp());

    await tester.tap(find.byKey(Key('plus')));
    await tester.pump();
    await tester.tap(find.byKey(Key('plus')));
    await tester.pump();
    await tester.tap(find.byKey(Key('plus')));
    await tester.pump();
    await tester.tap(find.byKey(Key('plus')));
    await tester.pump();
    await tester.tap(find.byKey(Key('plus')));
    await tester.pump();

    final c0 = find.text('0');
    expect(c0, findsOneWidget);
    final c3 = find.text('3');
    expect(c3, findsOneWidget);
    // No.4 is on the next page.
    var c4 = find.text('4');
    expect(c4, findsNothing);

    FlatButton button = find.widgetWithText(FlatButton, 'Delete 1').evaluate().first.widget as FlatButton;
    button.onPressed!();
    await tester.pump();

    c4 = find.text('4');
    expect(c4, findsOneWidget);

  });
}


lib/animated_grid.dart
library animated_grid;
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

/// Represents the order in which cells are placed in the grid
enum SortOrder {
  /// [1][2][3]
  /// [4][5]6]
  rightToLeft,
  /// [1][3][5]
  /// [2][4][6]
  topToBottom,
}

typedef GridCellWidgetBuilder = Widget Function(BuildContext context, int index, Size size);

/// Provides a grid view with animation.
///
/// Pass [keys] that uniquely identifies the item to be placed in the grid.
/// Newly added [keys] appear with the animation and disappear with the animation when deleted
class AnimatedGrid extends StatefulWidget {
  AnimatedGrid({
    required this.keys,
    required this.builder,
    this.width,
    this.height,
    this.cellColNum = 4,
    this.cellRowNum = 2,
    this.perCellMargin = 3,
    this.sortOrder = SortOrder.rightToLeft,
    this.scrollDirection = Axis.horizontal,
    this.pageController,
    this.onPageChanged,
  });

  final List<Object> keys;
  final GridCellWidgetBuilder builder;
  final double? width;
  final double? height;
  final int cellRowNum;
  final int cellColNum;
  final double perCellMargin;
  final SortOrder sortOrder;
  final Axis scrollDirection;
  final Function(int)? onPageChanged;
  final PageController? pageController;

  @override
  _AnimatedGridState createState() => _AnimatedGridState();
}

class _AnimatedGridState extends State<AnimatedGrid> {
  PageController? _pageController;
  final _dimensions = <Rect>[];
  int get cellNum => widget.cellRowNum * widget.cellColNum;
  var _prevKeyHashes = <int>[];

  @override
  void initState() {
    super.initState();

    switch(widget.sortOrder) {
      case SortOrder.rightToLeft:
        _createDimensionLeftToRight();
        break;
      case SortOrder.topToBottom:
        _createDimensionTopToBottom();
        break;
    }
  }

  void _createDimensionLeftToRight() {
    final gridW = widget.width ?? MediaQuery.of(context).size.width;
    final gridH = widget.height ?? MediaQuery.of(context).size.height;
    final w = (gridW - widget.perCellMargin) / widget.cellColNum - widget.perCellMargin;
    final h = (gridH - widget.perCellMargin) / widget.cellRowNum - widget.perCellMargin;
    double xPos = widget.perCellMargin;
    double yPos = widget.perCellMargin;
    for(var y=0; y<widget.cellRowNum; y++) {
      for(var x=0; x<widget.cellColNum; x++) {
        _dimensions.add(Rect.fromLTWH(xPos, yPos, w, h));
        xPos += w + widget.perCellMargin;
      }
      xPos = widget.perCellMargin;
      yPos += h + widget.perCellMargin;
    }
  }

  void _createDimensionTopToBottom() {
    final gridW = widget.width ?? MediaQuery.of(context).size.width;
    final gridH = widget.height ?? MediaQuery.of(context).size.height;
    final w = (gridW - widget.perCellMargin) / widget.cellColNum - widget.perCellMargin;
    final h = (gridH - widget.perCellMargin) / widget.cellRowNum - widget.perCellMargin;
    double xPos = widget.perCellMargin;
    double yPos = widget.perCellMargin;
    for(var x=0; x<widget.cellColNum; x++) {
      for(var y=0; y<widget.cellRowNum; y++) {
        _dimensions.add(Rect.fromLTWH(xPos, yPos, w, h));
        yPos += h + widget.perCellMargin;
      }
      xPos += w + widget.perCellMargin;
      yPos = widget.perCellMargin;
    }
  }

  @override
  Widget build(BuildContext context) {
    if(widget.pageController == null)
      _pageController = PageController();
    else
      _pageController = widget.pageController;

    return PageView.builder(
      scrollDirection: widget.scrollDirection,
      controller: _pageController,
      onPageChanged: (page) {
        if(widget.onPageChanged != null)
          widget.onPageChanged!(page);
      },
      itemBuilder: (ctx, index) {
        final st = index * cellNum;
        final ed = min(widget.keys.length, st + cellNum);
        final keys = widget.keys.sublist(st, ed);

        var _enableSlideIn = false;
        if(_prevKeyHashes.length != widget.keys.length) {
          _enableSlideIn = true;
        } else {
          for(var h in _prevKeyHashes) {
            var found = false;
            for(var k in widget.keys) {
              if(k.hashCode == h) {
                found = true;
                break;
              }
            }
            if(!found) {
              _enableSlideIn = true;
              break;
            }
          }
        }

        _prevKeyHashes.clear();
        for(var k in widget.keys) {
          _prevKeyHashes.add(k.hashCode);
        }

        return SizedBox.expand(
          child: Container(
            padding: EdgeInsets.only(left: widget.perCellMargin, top: widget.perCellMargin),
            child: PageGridBoard(dimensions: _dimensions, keys: keys, page: index, cellNum: cellNum, builder: widget.builder, enableSlideIn: _enableSlideIn,),
          ),
        );
      },
      itemCount: (widget.keys.length / cellNum).ceil(),
    );
  }
}

class PageGridBoard extends StatefulWidget {
  final List<Rect> dimensions;
  final List<Object> keys;
  final int page;
  final int cellNum;
  final GridCellWidgetBuilder builder;
  final bool? enableSlideIn;

  const PageGridBoard({required this.dimensions, required this.keys, required this.page, required this.cellNum, required this.builder, this.enableSlideIn});

  @override
  _PageGridBoardState createState() => _PageGridBoardState();
}

class _PageGridBoardState extends State<PageGridBoard> {
  final _cells = <Widget>[];
  final _cellKeys = <GlobalKey?>[];
  final _currentCellsMap = <Object, Widget>{};

  @override
  Widget build(BuildContext context) {
    for(var i=0; i<widget.keys.length; i++) {
      if(!_currentCellsMap.containsKey(widget.keys[i])) {
        final child = widget.builder(context, widget.page * widget.cellNum + i, widget.dimensions[i].size);
        final cell = PositionedCell(
          key: GlobalKey(),
          allDimensions: widget.dimensions,
          targetKey: widget.keys[i],
          allKeys: widget.keys,
          child: child,
          enableSlideIn: widget.enableSlideIn,
          onDeleted: (key) {
            final cell = _currentCellsMap.remove(key)!;
            _cellKeys.remove(cell.key);
            _cells.remove(cell);
          },
        );
        _cells.add(cell);
        _cellKeys.add(cell.key as GlobalKey<State<StatefulWidget>>?);
        _currentCellsMap[widget.keys[i]] = cell;
      }
    }
    for(var k in _cellKeys) {
      if(k!.currentState != null) {
        (k.currentState as _PositionedCellState).updateKeys(widget.keys);
      }
    }

    return Stack(
      children: _cells,
    );
  }
}

class PositionedCell extends StatefulWidget {
  final List<Rect> allDimensions;
  final Object? targetKey;
  final List<Object>? allKeys;
  final Widget? child;
  final Function(Object?)? onDeleted;
  final bool? enableSlideIn;

  const PositionedCell({Key? key, required this.allDimensions, this.targetKey, this.allKeys, this.child, this.onDeleted, this.enableSlideIn}) : super(key: key);

  @override
  _PositionedCellState createState() => _PositionedCellState(targetKey, allKeys);
}

class _PositionedCellState extends State<PositionedCell> with SingleTickerProviderStateMixin {
  _PositionedCellState(this._targetKey, this._allKeys);

  var _deleted = false;
  Rect _prevRect = Rect.zero;
  Object? _targetKey;
  List<Object>? _allKeys;

  late AnimationController _animationController;
  late Animation<Offset> _animation;

  Object? get tagetKey => _targetKey;

  @override
  void initState() {
    super.initState();
    _animationController = AnimationController(vsync: this, duration: const Duration(milliseconds: 300))
      ..addListener(() {setState((){});});
    _animation = Tween<Offset>(begin: const Offset(0, 2), end: Offset.zero).animate(_animationController);
    _animationController.forward();
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  void updateKeys(List<Object> keys) {
    setState(() {
      _allKeys = keys;
    });
  }

  @override
  Widget build(BuildContext context) {
    Rect? rect;
    for(var i=0; i<_allKeys!.length; i++) {
      if(_allKeys![i] == _targetKey) {
        rect = widget.allDimensions[i];
        break;
      }
    }

    // deleted.
    if(rect == null) {
      rect = Rect.fromLTWH(_prevRect.left, -600, _prevRect.width, _prevRect.height);
      _deleted = true;
    } else {
      _prevRect = rect;
    }

    final child = widget.enableSlideIn! ?
        SlideTransition(
          position: _animation,
          child: widget.child,
        ) : widget.child! ;

    return AnimatedPositioned(
      left: rect.left,
      top: rect.top,
      width: rect.width,
      height: rect.height,
      duration: const Duration(milliseconds: 300),
      child: child,
      onEnd: () {
        if(_deleted)
          widget.onDeleted!(widget.targetKey);
      },
    );
  }
}

example/animated_grid_example/lib/main.dart
import 'package:animated_grid/animated_grid.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(ExampleApp());
}

class ExampleApp extends StatefulWidget {
  @override
  _ExampleAppState createState() => _ExampleAppState();
}

class _ExampleAppState extends State<ExampleApp> {
  final _keys = <int>[];
  var _count = 0;
  final _pageController = PageController();

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      color: Colors.black,
      home: Builder(
        builder: (context) {
          return Scaffold(
              appBar: AppBar(
                title: Text('Example'),
                actions: [
                  IconButton(
                    icon: Icon(Icons.add),
                    onPressed: () {
                      setState(() {
                        _keys.add(_count);
                        _count++;
                      });
                    },
                  ),
                  IconButton(
                    icon: Icon(Icons.chevron_left),
                    onPressed: () {
                      _pageController.previousPage(duration: const Duration(milliseconds: 300), curve: Curves.easeIn);
                    },
                  ),
                  IconButton(
                    icon: Icon(Icons.chevron_right),
                    onPressed: () {
                      _pageController.nextPage(duration: const Duration(milliseconds: 300), curve: Curves.easeIn);
                    },
                  ),
                ]
              ),
              backgroundColor: Colors.black,
              body: AnimatedGrid(
                keys: _keys,
                width: MediaQuery.of(context).size.width,
                height: MediaQuery.of(context).size.height - 44,
                cellRowNum: 2,
                cellColNum: 4,
                pageController: _pageController,
                sortOrder: SortOrder.rightToLeft,
                scrollDirection: Axis.horizontal,
                onPageChanged: (page) {},
                builder: (ctx, index, _) {
                  return ExampleContent(
                    caption: '${_keys[index]}',
                    keyVal: _keys[index],
                    onDeleteAt: (key) {
                      setState(() {
                        _keys.remove(key);
                      });
                    },
                  );
                },
              ),
          );
        },
      ),
    );
  }
}

class ExampleContent extends StatelessWidget {
  const ExampleContent({Key key, this.caption, this.keyVal, this.onDeleteAt}) : super(key: key);

  final String caption;
  final int keyVal;
  final Function(int) onDeleteAt;

  @override
  Widget build(BuildContext context) {
    return Container(
      color: Colors.white,
      child: Column(
        children: [
          const Spacer(),
          Text(caption),
          const SizedBox(height: 20),
          FlatButton(
            onPressed: () {
              onDeleteAt(keyVal);
            },
            child: Text('Delete'),
          ),
          const Spacer(),
        ],
      ),
    );
  }
}